### DISCUSSION TITLE
Arrow filters vs code filters and SpineInterface

### DISCUSSION THREAD

**Discussion by DillonJ** [Spine-project/Spine-Toolbox#2377] [Support]
I wonder could someone clarify the behaviour of Julia and Python tools that directly access SpineDBs using SpineDB_api

If I specify a scenario filter in an arrow from a SpineDB to a Julia tool for example, and then the Julia tool uses SpineInterface and specifies the `filters` argument in the call to `using_spinedb()` are both filters applied? In what order?

My understanding is that the url argument the tool sees has the filter embedded... so SpineInterface sees a url with a filter embedded and also an explicit "filters" argument. What does it do?

What are the rules?

If I specify a scenario filter in the arrow to the tool and don't specify the filters argument in the call to `using_spinedb()` I presume the arrow filter is applied as it's in the url?

If I specify no scenario filter in the arrow but do supply the filters argument, I presume this filter is applied?

Thanks in advance

**Comment by soininen** [comment]
I took the liberty of converting this to a discussion as it looks more like that, not an issue.

**Comment by manuelma** [comment]
I would suggest not to mix the two ways of specifying filters.

The reason is, if you specify the filters argument to using_spinedb, all filters will be cleared afterwards (including those specified in the arrow in Spine Toolbox).

This is so each call to using_spinedb leaves the DB unfiltered, so the next call can apply their own filters. Ideally, each call to using_spinedb should leave the DB in the state it found it (say, with arrow filters from Spine Toolbox) but that's not implemented.

**Reply by DillonJ** [reply]
@manuelma - understood, but could you confirm that if an arrow filter is applied in toolbox, that the call to `using_spindb()` sees this filter? 

Could you also please clarify what happens if both arrow filter and filter argument are used - I understand this shouldn't be done

**Reply by manuelma** [reply]
> could you confirm that if an arrow filter is applied in toolbox, that the call to using_spindb() sees this filter?

Yes, using_spinedb() without specifying the filters keyword argument won't touch the current filter structure, so the ones from toolbox arrows will prevail.

**Reply by manuelma** [reply]
> what happens if both arrow filter and filter argument are used

Funny things in general, but if you call using_spinedb only once, then arrow filter and using_spinedb filter will be applied in sequence as long as they are not incompatible (e.g. selecting scenario A on the arrow and then filtering by scenario B in using_spinedb, this won't work).

The second time you call using_spinedb, all filter will be ignored (both arrow's and using_spinedb's)
